### ISSUE TITLE
VideoPlayerController.memory feature Request

### ISSUE THREAD

**Issue by Loopex2019** [flutter/flutter#53017]
Can someone play videos from Memory : something Like :  `VideoPlayerController.memory(Unit8List)`.
There are already the features `.file()` , `.asset()` , `.network()` . Maybe you can add the `.memory()` constructor .

**Comment by theskyblockman** [comment]
There's a package named just_audio who allows you to read an audio track from a **stream** of bytes, to do this they used a local server that transfers data between the two ends. I think something like this could be implemented in the package in its current state, although the data live decoding could be a technical challenge to overcome.

**Comment by MarcSoft54** [comment]
please can we play video from Uint8List?

**Comment by MarcSoft54** [comment]
i want for the mobile phone

**Comment by theskyblockman** [comment]
> please can we play video from Uint8List?

Actually this isn't really viable for videos, they store too much memory for a mobile phone, a full buffering system is needed for something like this to be performant.

**Comment by MarcSoft54** [comment]
Ah okays and please how can I do to plays video from Uint8List

Please I want help

**Comment by theskyblockman** [comment]
> Ah okays and please how can I do to plays video from Uint8List Please I want help

There's a a plugin named [BetterPlayer](https://pub.dev/packages/better_player) which is likely not maintained anymore. Someone made a fork and works on it, although you're going to know how to install a package from git in your project. Here's the Github repository: https://github.com/AnonymHK/betterplayer

**Comment by MarcSoft54** [comment]
Thank you very much

**Comment by MarcSoft54** [comment]
morning Flutter/Flutter
i have read you package very well and went i have implement it
on android the thing is giving me this error message and on windows i have
another message

i will send you the message

FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring project ':better_player'.
> Could not resolve all files for configuration ':better_player:classpath'.
   > Could not download kotlin-compiler-embeddable-1.5.31.jar
(org.jetbrains.kotlin:kotlin-compiler-embeddable:1.5.31)
      > Could not get resource '
https://repo.maven.apache.org/maven2/org/jetbrains/kotlin/kotlin-compiler-embeddable/1.5.31/kotlin-compiler-embeddable-1.5.31.jar
'.
         > Could not GET '
https://repo.maven.apache.org/maven2/org/jetbrains/kotlin/kotlin-compiler-embeddable/1.5.31/kotlin-compiler-embeddable-1.5.31.jar
'.
            > H�te inconnu (repo.maven.apache.org)
> Failed to notify project evaluation listener.
   > Could not get unknown property 'android' for project ':better_player'
of type org.gradle.api.Project.
   > Could not find method implementation() for arguments [project
':wakelock'] on object of type
org.gradle.api.internal.artifacts.dsl.dependencies.DefaultDependencyHandler.
   > Could not get unknown property 'android' for project ':better_player'
of type org.gradle.api.Project.


*this is the error message for the android*




[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception:
MissingPluginException(No implementation found for method init on channel
better_player_channel)
#0      MethodChannel._invokeMethod
(package:flutter/src/services/platform_channel.dart:308:7)
<asynchronous suspension>

[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception:
MissingPluginException(No implementation found for method create on channel
better_player_channel)
#0      MethodChannel._invokeMethod
(package:flutter/src/services/platform_channel.dart:308:7)
<asynchronous suspension>
#1      MethodChannelVideoPlayer.create
(package:better_player/src/video_player/method_channel_video_player.dart:37:37)
<asynchronous suspension>
#2      VideoPlayerController._create
(package:better_player/src/video_player/video_player.dart:202:18)
<asynchronous suspension>


and this is for the desktop (windows)

Le jeu. 21 sept. 2023 à 18:43, Marc William ***@***.***> a
écrit :

**Comment by MarcSoft54** [comment]
am not understand this message

Le ven. 22 sept. 2023 à 10:30, Marc William ***@***.***> a
écrit :

**Comment by MarcSoft54** [comment]
i have install the github repository but the thing is know showing another
error about the first message that i have show you first


Le ven. 22 sept. 2023 à 10:31, Marc William ***@***.***> a
écrit :

**Comment by mnayef95** [comment]
Any update on this?

**Comment by theskyblockman** [comment]
> Any update on this?

No, currently I don't think there's any way to play a video entirely from memory in Flutter. You can save a file then play it from the local storage but you can't skip the local storage for the time being.

**Comment by theskyblockman** [comment]
I also want to clarify that loading a full video into memory will mostly result into memory overflow, a simple ``.memory()`` is impossible. A caching/buffering system must be made, which could mean a direct communication with the platforms player (ex. ExoPlayer in Android) would be required and would need to be quick enough to get good performance. Currently every single way to play a video just send the source of the content to be played not the content itself.

But I looked a little bit into the [Audioplayers](https://pub.dev/packages/audioplayers)'s source code and they seem to send audio data through a local server which could (*maybe...*) be doable with good enough performance. Technically the communication is only on the local machine, but it seems like the INTERNET permission on Android is still required but you don't need to be online. This could still be a viable way to play videos kinda from memory.

**Comment by MarcSoft54** [comment]
Thank you

Le lun. 4 déc. 2023, 21:31, El Omri Haroun ***@***.***> a
écrit :

**Comment by MarcSoft54** [comment]
I will try it

Le dim. 26 nov. 2023, 18:24, El Omri Haroun ***@***.***> a
écrit :

**Comment by GleammerRay** [comment]
@theskyblockman it fully depends on the video in question. I am designing my application with video size of under 50Mb in mind, so loading them into memory does not pose any problems. Having a `.memory()` constructor would be extremely useful.

**Comment by theskyblockman** [comment]
@GleammerRay In this case ``.memory()`` could be useful, but I am still against this because:
- I am always worried of [the overhead made by moving data from Flutter to Host](https://docs.google.com/document/d/1bD_tiN987fWEPtw7tjXHzqZVg_g9H95IS32Cm609VZ8/edit) (to be clear, this document shows small bits of data being transferred, only tests for mobile and is 4 years old but shows that at best it is 2-3x slower using JNI which is still not really practical today)
- Developers will use this for their 4 GB video, see their app crash, blame Flutter and then complain to have a way to stream the data. Also if they have small videos they could duplicate the data while they are handling it and keep unused pointers to it

Generally I find that possible user error is not a valid reason to be against something but here I feel like we would only give a bad way to do things without any other options to do it.

If we ever do video playing from memory I think we must rather ask for an object the user implements which enables us to get parts of the data to make the video player seekable and to not make the Flutter/Native side of the memory explode. We would also require some metadata to setup the player (mime type, bitrate, etc...). By doing it this way the user is way more explicit about keeping the data in memory.

**Comment by helomri** [comment]
> @helomri it can be useful for small clips, live avatars or etc
> developer cares about memory and resources based on a certain case

It can be useful in those cases (and other ones mentioned above) and most if not all people wanting to use something like this wouldn't have any issue using it for their legitimate applications, but for those that take really little consideration into their app's memory budget (who I see more and more of) it will certainly be a big problem. If there was a smart and efficient way to ensure only those legitimate uses of this constructor, I wouldn't be opposed to this, only this is afaik not really feasible. It could make such devs comfortable using video players with videos they can run on their device but not on their users's. Forcing them to make the data available through Random Access is the safest way to do this. I don't think the official Flutter package should have this ``.memory()`` constructor, but if someone wants to create a little extension that would do that, they are welcome to do so (after/if anyone ever starts working on this, which is unlikely since I am not available for this anytime soon and the Flutter team is busy doing many other things).

**Comment by zs-dima** [comment]
@helomri this is too special a case. Developer knows the certain cases, and if there is a chance for memory budget or etc, then the developer will take it into account. It is a bad practice when plugin is limiting developers instead of provide them possibilities for certain cases.
I know many cases when a very-short-video is already in memory as a Uint8List, but during video_player limitation, we have to save and load it as a file once more.

**Comment by helomri** [comment]
> @helomri this is too special a case. Developer knows the certain cases, and if there is a chance for memory budget or etc, then the developer will take it into account. It is a bad practice when plugin is limiting developers instead of provide them possibilities for certain cases.
> I know many cases when a very-short-video is already in memory as a Uint8List, but during video_player limitation, we have to save and load it as a file once more.

I think devs should be able to read videos from `UInt8List``s, what I mean to say is to make it so it is as easy to read the video from a list than it is to load it progressively. ``.memory()`` would mainly mean to me that Flutter recommends you to play a video from memory (and .memory() would simply be some kind of helper method implementing the progressive video player but keep the whole file inside it).

What I am trying to say is to incite the dev as much as possible to not keep the whole video in memory but to not totally constrain them to it (which is probably impossible anyway)
